### ISSUE TITLE
Logitech G502 X Plus

### ISSUE THREAD

**Issue by uggwar** [libratbag/libratbag#1344]
Hey!

Trying to get basic support for my G502 X Plus, but I'm running into some trouble. I am able to list the USB device id for the receiver, but the mouse itself isn't showing up. My old G305 works great, but the id in the .device file is nowhere to be seen, so I obviously lack some knowledge here. :-) Please help me along!

**Comment by Vandool** [comment]
Hi, I have the same issue, `lsusb` lists my `Logitech G502 X PLUS` when wired, but won't when wireless.

`ratbagd --verbose | grep G502` returns:
```
ratbag debug: New device: Logitech G502 X PLUS
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw6'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw7'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw8'.
ratbag debug: 'Logitech G502 X PLUS' is using protocol v4.2
ratbag raw: 'Logitech G502 X PLUS' has 33 features
ratbag error: Profile layout not supported: 0x05.
ratbag error: Logitech G502 X PLUS: error opening hidraw node (Operation not supported)
ratbag debug: New device: Logitech G502 X PLUS
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw6'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw7'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw8'.
ratbag debug: 'Logitech G502 X PLUS' is using protocol v4.2
ratbag raw: 'Logitech G502 X PLUS' has 33 features
ratbag error: Profile layout not supported: 0x05.
ratbag error: Logitech G502 X PLUS: error opening hidraw node (Operation not supported)
ratbag debug: New device: Logitech G502 X PLUS
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw6'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw7'.
ratbag debug: Logitech G502 X PLUS is device '/dev/hidraw8'.
ratbag debug: 'Logitech G502 X PLUS' is using protocol v4.2
ratbag raw: 'Logitech G502 X PLUS' has 33 features
ratbag error: Profile layout not supported: 0x05.
ratbag error: Logitech G502 X PLUS: error opening hidraw node (Operation not supported)
```
Any idea how I might get this thing working? Thanks in advance!

**Comment by ericshenshs** [comment]
Same issue here.

**Comment by doogie544** [comment]
I can verify this as well.  I will not post a log as Vandool already did and mine is identical.  I can also verify that it does not even find the mouse if connected wirelessly, but this is expected as this is not supported as mentioned here https://github.com/libratbag/piper/issues/388#issuecomment-1253901281


One thing I can mention that might help narrow things down.  In Windows this mouse is also not supported but the Logitech Gaming Software, it will not be found at all, but instead needs the Logitech G Hub to be supported.  Could that mean that it uses a different driver so `Driver=hidpp20` is not the driver that supports this mouse?  Maybe there is a different driver that is available that ratbag can use, or maybe there is not one yet written for linux that supports whatever protocol is used by this mouse?

**Comment by JonWel** [comment]
Same issue with the "Logitech G502 X" (so not the "Plus" version, but main difference around this issue is that I'm not concerned by RGB), connected though PowerPlay Wireless Charging System (the mat that allows you to charge the mouse while using, and behave like a receiver).
`lsusb` only list the PowerPlay mat:
`Bus 001 Device 003: ID 046d:c53a Logitech, Inc. PowerPlay Wireless Charging System`

Left clic, right click and vertical scrolling works.
Hope to use the other functions too.

**Comment by Seb-Solon** [comment]
Hi there,


Tried to recompile current master branch with the following patch :

```C
@@ -1802,6 +1802,7 @@ int hidpp20_adjustable_report_rate_set_report_rate(struct hidpp20_device *device
 #define HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G303     0x02
 #define HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G900     0x03
 #define HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G915     0x04
+#define HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G502     0x05
 #define HIDPP20_ONBOARD_PROFILES_MACRO_TYPE_G402       0x01
 
 #define HIDPP20_USER_PROFILES_G402                     0x0000
@@ -2229,7 +2230,8 @@ hidpp20_onboard_profiles_validate(struct hidpp20_device *device,
        if ((info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G402) &&
            (info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G303) &&
            (info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G900) &&
-           (info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G915)) {
+           (info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G915) &&
+           (info->profile_format_id != HIDPP20_ONBOARD_PROFILES_PROFILE_TYPE_G502)) {
                hidpp_log_error(&device->base,
                                "Profile layout not supported: 0x%02x.\n",
                                info->profile_format_id);

```

The error is removed with this and `ratbagctl list` returns `chanting-squirrel:   Logitech G502 X PLUS ` when plugged through USB
However, if I try to do any command (changing color, setting profile), the led of the mouse start getting "crazy" :  fast rainbow effect and nothing can make it stop. 
The only way to reset to default is to create a new profile and do not try to change anything

I feel like the underlying protocol is somehow different so the command send to the mouse are mis-interpreted. 

@staticssleever668 Any take on it ? Not sure if there is new version of hid++ or Logitech just decided to add proprietary stuff on top of it.

**Comment by staticssleever668** [comment]
Sorry, I have never worked with HID++ nor did I ever have a Logitech device to even try it... We would have to wait until someone with better knowledge about it comes up.

> Not sure if there is new version of hid++ or Logitech just decided to add proprietary stuff on top of it.

As far as I know HID++ was made *by* Logitech in the first place, so I don't think it's fair to phrase this in such a way. Probably just some new features we don't support yet.

---

By the way, has anyone tried yet if it works with [Solaar](https://github.com/pwr-Solaar/Solaar)?

**Comment by Seb-Solon** [comment]
Well I am not very familiar with the protocol too. So yeah, I believe they have a new version of the protocol for that new mouse (out in Aug 2022).

I gave a try with Solaar but I think it is not support yet neither. I tried to add support by copying G502 Hero but getting a low level error. 

The best is to wait or manage to get more info on the protocol. 

Thanks for the quick reply :)

**Comment by JonWel** [comment]
I made some experiment.
If I plug the USB cable, my `lsusb` is: `Bus 001 Device 023: ID 046d:c098 Logitech, Inc. G502 X LIGHTSPEED`, which has one digit different than the "Plus" version (I have "Logitech G502 X").

Given the suggestion made by @staticssleever668 , I created a file `/usr/share/libratbag/logitech-g502-x-wireless.device`, with:
```
[Device]
Name=Logitech G502 X
DeviceMatch=usb:046d:c098
Driver=hidpp20
```

Then `sudo systemctl restart ratbagd.service`, and now `ratbagctl list` gives:
`screaming-acouchy:   Logitech G502 X LIGHTSPEED`

And I can select the mouse in Piper (only when wired).
I don't have a nice graphic showing where the buttons are, changing the DPI works (when clicking in the menu).
On the button range, I get only 8 of them, and I don't seem to find the one to change the DPI on the fly.
If I disconnect the USB cable, DPI settings go back to default, and aren't coming back if I plug the cable again.

**Comment by zagtastic** [comment]
i have gotten a similar result as JonWel describes above (partial piper support) with a logitech g502 x plus connected via lightspeed mousepad by adding a .device file with:

`[Device]
Name=Logitech G502 X+ Lightspeed
DeviceMatch=usb:046d:409f
Driver=hidpp20`

note:  both my piper and libratbag are built from latest source.

**Comment by RobinFrcd** [comment]
Hi, I have the same issue on my G502 X PLUS, either with or without a USB cable.

Is there any plan to support this mouse ? Or any quick fix to at least be able to change the keybinds on the buttons ? 

Thanks !

**Comment by Riven-Spell** [comment]
Yep, I'm here too.

**Comment by Riven-Spell** [comment]
*for now* until libratbag supports the mouse, your best option is to take it to a Windows computer (or hook it up to a Windows VM), configure the keys to all press a real keyboard button or whatnot, set your DPI preferences, apply that profile to the onboard memory, take it back to your Linux machine and use something like [input remapper](https://github.com/sezanzeb/input-remapper) to adjust keys on the fly.

**Comment by doogie544** [comment]
One thing you guys reporting need to consider.  libratbag may not ever support this mouse.  I may need to be completely rewritten to allow that:

As I mentioned earlier this mouse does not even work on Windows with the Logitech Gaming Software, LGS does not find the mouse either.  I would assume that libratbag uses the same method LGS does to detect the mice.  

The only way to setup the mouse on windows is the use the Logitech G Hub which obviously uses a different protocol to communicate with the mice.  libratbag may need to be completely rewritten to use the new protocol.  Not only is this likely a large task for the devs of it to undertake, it is also possible that in doing so they will lose compatibility with a large portion of the mice they can now support.  

They may have to write it as a separate program (libratbag2 for example)  to be compatible with the newer protocol without losing compatibility with the older mice, which would include non gaming mice as well.  If that is what has to happen then Piper is probably going to have to be rewritten to be able to use both protocols or will have to also come out with a newer program just for the newer protocol (Piper-Gaming for example)

The devs of piper and libratbag are aware of the issue but who know how quick, if ever, they will be able to support these mice or the new protocol that Logitech G Hub uses

**Comment by jnvsor** [comment]
> may need to be completely rewritten

Doubt that. libratbag already supports multiple protocols. A new protocol would have to be written but it's not on the level of rewriting the entire library... Would have to be reverse engineered though, and no-one seems to have time for that

**Comment by doogie544** [comment]
> > may need to be completely rewritten
> 
> Doubt that. libratbag already supports multiple protocols. A new protocol would have to be written but it's not on the level of rewriting the entire library... Would have to be reverse engineered though, and no-one seems to have time for that


well that is good news.  Still it may be a while before that is able to be reverse engineered and added.

Still  even taking the correction into account the overall point stands that it may be while so patience is needed here

**Comment by gibtsnet** [comment]
> Hi, I have the same issue on my G502 X PLUS, either with or without a USB cable.
> 
> Is there any plan to support this mouse ? Or any quick fix to at least be able to change the keybinds on the buttons ?
> 
> Thanks !

I have the cable only version and was able to change the keybinds via console. for example: 
ratbagctl 'Logitech G502 X' button 6 action set macro KEY_F11
ratbagctl 'Logitech G502 X' button 9 action set macro KEY_APOSTROPHE
ratbagctl 'Logitech G502 X' button 4 action set button 5

Hope that helps. I could not change the LED behavior.

**Comment by patofet** [comment]
G502 X here, remaping buttons doesn't work at all I supose that is normal

**Comment by gibtsnet** [comment]
> G502 X here, remaping buttons doesn't work at all I supose that is normal

I have the cable version. It worked via console. e.g.:
ratbagctl 'Logitech G502 X' button 7 action set macro KEY_APOSTROPHE
ratbagctl 'Logitech G502 X' button 9 action set macro KEY_F9
ratbagctl 'Logitech G502 X' button 4 action set button 5

**Comment by jnvsor** [comment]
Bought a G502 X after @gibtsnet comment.

Piper works for DPI, but button assignment doesn't work and it only detects 8 buttons.

ratbagctl works for DPI and buttons BUT all the mouse profiles are disabled by default. Until you enable one with `ratbagctl 'Logitech G502 X' profile $n enable` the commands won't have any visible effect

tl;dr: Confirm CLI works but it's a bit trickier than GUI obviously

**Comment by YamiYukiSenpai** [comment]
The G502 X & G502 X Plus are 2 different devices

**Comment by d0x360** [comment]
I can say that the new 502 hero x plus doesn't need GHUB to function.

GHUB broke both my 502 hero and 915 tkl in a may update.  They ran down the warranty with nonsense support messages that took a week + to arrive.

So even the side buttons work and I assume they are using drivers from win update because this is a clean install.

**Comment by doogie544** [comment]
> I can say that the new 502 hero x plus doesn't need GHUB to function.
> 
> GHUB broke both my 502 hero and 915 tkl in a may update. They ran down the warranty with nonsense support messages that took a week + to arrive.
> 
> So even the side buttons work and I assume they are using drivers from win update because this is a clean install.



It does not need the ghub software to *function*  it needs it to program the buttons to something other than the default functions.  i.e. you need it to change the on-board memory.  iirc you also need it to enable more than one profile for the on-board memory.

It will function with the default profile in  windows just like it will on Linux.  However to *change the button bindings*  you need to use Ghub.  It has always been the case that you needed software of some kind to change the buttons with the g502 mice regardless of the particular model name they had at the time.  What has changed is that up until the g502 X series you could use the Logitech Gaming Software in Windows as well as Ghub and that you could use Piper for the 502 up until the 502X.  

With the 502X Plus (and I think the other 2 models of the 502x)  what changed was that Logitech Gaming Software does not detect it, you *have* to use the Ghub software to find the mouse and change the buttons.  Once those buttons are saved to the memory you don't need Ghub to use the mouse, only if you want to change the buttons again.

**Comment by d0x360** [comment]
Ok I gotcha. Default was changed but the changes just removed functions..

**Comment by arnauddupuis** [comment]
> By the way, has anyone tried yet if it works with [Solaar](https://github.com/pwr-Solaar/Solaar)?
Just tried it: nothing when plugged through USB, however, in wireless mode you can see the mouse and change some values (like DPI or polling rate). No rgb or re-mapping the buttons though.

**Comment by KeebWasTaken** [comment]
Any updates on any of this? I created a device myself by changing this in the files `DeviceMatch=usb:046d:c095` but get this error when I unplug and re-plug the mouse in the terminal:
> profile layout not supported: 0x05

**Comment by CorvetteCole** [comment]
same here. I think support is a long way off unfortunately. I may return the mouse

> Any updates on any of this? I created a device myself by changing this in the files `DeviceMatch=usb:046d:c095` but get this error when I unplug and re-plug the mouse in the terminal:
> 
> > profile layout not supported: 0x05

**Comment by Taztaztaz** [comment]
I found that when I set settings in windows and activate onboard profiles, then SOLAAR shomewhat works

**Comment by monsieurpigeon** [comment]
in my case i have this line after a `lsusb`

```Bus 001 Device 005: ID 046d:c099 Logitech, Inc. G502 X```

And it kind of work when i use `DeviceMatch=usb:046d:c099`
The mouse is recognized, but I don't have access to all the buttons of the mouse currently, still searching

**Comment by paulo-juchimiuk** [comment]
Hello everyone, congratulations on the great work. However, could you clarify something for me? I've been following the issue, and I know we don't yet have support for the 502x plus. I just want to know if the mouse at least performs basic functions when connected via Bluetooth? I'm considering getting a mouse, and I think it's worth getting a current model and waiting for future Linux support updates. By the way, since I'll use the mouse for work on Linux and gaming on Windows, I'll try to collaborate with the community to get this new device compatible. But first, I need to know if I can at least work with it connected via Bluetooth. LOL.

**Comment by CorvetteCole** [comment]
Bluetooth? I don't think it even supports Bluetooth

**Comment by d0x360** [comment]
If you want a good mouse that supports wired, Bluetooth and 2.x GHz I
highly recommend the rog chakram x.  I have used the wireless 502 for years
and upgraded to the new model day 1 but then I got to use the chakram and
it's just so much better overall.

I don't think the Logitech is Bluetooth compatible at all but I could be
wrong.  I've had multiple high priced Logitech keyboards and mice die on me
and the last straw was when a ghub update (not even a firmware update)
somehow bricked the 502 and my 2nd g915.  They both immediately had
connection issues after the update and nothing would solve it.

I contacted support and both had around 4 months! Left on the warranty and
Logitech literally strung me along by sending me log requests and things to
try but their email responses would often take 2-3 weeks.

Right before the warranty expired they decided my USB receivers were dead
even though they weren't.  So they sent me 2 beat up used ones and had me
pair the devices and report back.  The day I did this there was 16 days
left on the 915 warranty and 22 on the 502.

On day 23 they finally replied and said that's all they can do.  I was
beyond mad and it's not like I couldn't just buy new ones and return the
busted ones although I don't ever want to be forced down that road I will
go down it if a company fails so completely.

I figured because the g915 was DOA and I had to get a warranty
replacement.  Now for all I knew the issue could have been the receiver but
they had me film myself destroying it and only then would they send one so
I did and I guess they were mad that a 20+ year customer that always bought
their latest and greatest things had to actually use the warranty for once.

I mean my profile showed a lot of what I had and that included a 913, a
powermat, 4 yes I had 4 of the newest wheels and pedals they made at the
time.  I also had over the years every high end Harmony except that one
that was like $2000 and square.  I even had multiple harmony repeaters that
were only sold with harmony one's.

They lost a whale and now I use a wooting keyboard and I have a couple mice
I use but mostly the rog one after testing everything on the market that
seemed half decent.

Now I have replaced everything but the harmony but once it dies it dies.  I
love my fanatec racing rig and I don't have 4 but I do have 3 wheels so I
can play on ps5, Xbox and PC and I needed a PC/Xbox wheel that would work
for rally and GT racing and another that was F1 style.

Sorry for the rant...topic brought back some heat lol.

**Comment by doogie544** [comment]
> Hello everyone, congratulations on the great work. However, could you clarify something for me? I've been following the issue, and I know we don't yet have support for the 502x plus. I just want to know if the mouse at least performs basic functions when connected via Bluetooth? I'm considering getting a mouse, and I think it's worth getting a current model and waiting for future Linux support updates. By the way, since I'll use the mouse for work on Linux and gaming on Windows, I'll try to collaborate with the community to get this new device compatible. But first, I need to know if I can at least work with it connected via Bluetooth. LOL.

The mouse does not have bluetooth support but if you are asking if it works with the default buttons.  The answer is short of.

The left & right click work, middle mouse button works completely (side to side scrolling, mouse wheel and middle click), the 2 thumb buttons work (the dpi reduction button may work as well I can't recall for sure).  What does not work b/c it does not have any output by default are the 2 extra buttons by the left mouse button.

If you use the logitech ghub software in windows or a windows vm then you can save the setting to the onboard memory.  The extra left buttons still will not produce any output by default (none of the 502s did iirc)  but you can set it up as keyboard input for a key that is not used in games typically (I use numpad key assigments but they could be anything)  and then in game set the binding for the action you want with that mouse button, the game will just think you are using that key and save it in game as such

**Comment by Thomas-git** [comment]
@paulo-juchimiuk I used another computer on windows to setup the mouse and saved params onboard. All buttons works on linux (including dpi, profil switching and lateral buttons). Same goes for RGB setup.

**Comment by paulo-juchimiuk** [comment]
First, I'd like to thank you for the answers; they cleared up many of my doubts. And indeed, I was convinced that this model had Bluetooth, @CorvetteCole thanks for pointing that out. @d0x360, I researched that mouse, and it indeed covers everything I need for work and gaming. However, it's not available for sale here in Brazil, and I'd have to import it. The import taxes in Brazil are 60% of the product's value, which is insane, unfortunately. @doogie544 @Thomas-git, good to know, friends; that's a significant point I wasn't aware of and will be of great help. Anyway, thanks everyone for the assistance!

**Comment by d0x360** [comment]
Ouch yeah I know all about your import taxes because when I was a wee lad I
worked at GameStop and we always had 2 groups trying to buy all of console
a or b so we had to limit them because the store didn't really profit on
them and they never bought games or accessories and for every console we
sold them it was a console that we couldn't sell to a likely regular
shopper who got games.

I asked one of them one day and they told me about how much tax gets put on
a product if it's not made in Brazil and I was shocked because usually you
only see numbers that high to punish another country then get rid of the
tax when the issue was resolved.

The other group was buying them to sell in China because they were illegal
at that point.

I had some time to pass so I figured I'd burn some of it with a warning
that Logitech of today isn't quite as good as they used to be.  Also before
using any version of ghub make sure it didn't cause issues because it's
something that was becoming more common before I moved on.  Ghub caused
things to stop working in 3/10 updates and the hardware was fine after they
pushed a fixed update usually in 6 to 24 hours...

I gotta start sleeping lol.  Good luck man.  The mouse is good it's just
their software that seems to get worse and worse but that's a bigger
concern with keyboards than it is mice.

**Comment by tbelway** [comment]
I have this device and thus if you need something in addition let me know.

**Comment by xQuandaleDinglex** [comment]
Hi, I'm bumping this thread up if it helps. Libratbag doesn't seem to have support for the G502 X Plus as of 4/12/24. :/

**Comment by Goku-San** [comment]
Hi, I am writing just to get notified if this is fixed. I have G502 X Plus wireless...

**Comment by senorYME** [comment]
yep still not finding on g502 x plus wireless

**Comment by Redcarred2** [comment]
Someone needs to put in the effort, this issue will be marked as closed as soon as it's fixed

**Comment by Thomas-git** [comment]
I can't help but I want to congratulate you for this major step toward G502 support.

**Comment by MOVZX** [comment]
Setting the proper device index for the USB receiver works great.

```
$ ratbagctl singing-gundi info
singing-gundi - Logitech USB Receiver
             Model: usb:046d:c547:0
       Device Type: Mouse
 Number of Buttons: 11
    Number of Leds: 1
Number of Profiles: 5
Profile 0:
  Name: n/a
  Report Rate: 1000Hz
  Resolutions:
    0: 200dpi
    1: 400dpi
    2: 500dpi (active) (default)
    3: 800dpi
    4: 1000dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 5'
  Button: 4 is mapped to macro '↕KEY_PLAYPAUSE'
  Button: 5 is mapped to 'button 4'
  Button: 6 is mapped to macro '↕KEY_PREVIOUSSONG'
  Button: 7 is mapped to macro '↕KEY_NEXTSONG'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'button 1'
  Button: 10 is mapped to macro '↕KEY_MIN_INTERESTING'
  LED: 0, depth: rgb, mode: breathing, color: ff0000, duration: 5000, brightness: 142
Profile 1: (active)
  Name: n/a
  Report Rate: 1000Hz
  Resolutions:
    0: 200dpi
    1: 450dpi
    2: 500dpi (active) (default)
    3: 800dpi
    4: 1000dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 5'
  Button: 4 is mapped to macro '↕KEY_B'
  Button: 5 is mapped to 'button 4'
  Button: 6 is mapped to macro '↕KEY_F5'
  Button: 7 is mapped to macro '↕KEY_F3'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'button 2'
  Button: 10 is mapped to macro '↕KEY_F4'
  LED: 0, depth: rgb, mode: breathing, color: ff0000, duration: 3000, brightness: 140
Profile 2: (disabled)
Profile 3: (disabled)
Profile 4: (disabled)
```

**Comment by shaqaruden** [comment]
how has support for this mouse not been officially added yet? IT'S OVER TWO YEARS OLD

**Comment by CorvetteCole** [comment]
> how has support for this mouse not been officially added yet? IT'S OVER TWO YEARS OLD

This is an open source project run by volunteers. Having expectations with no money attached to them is quite disrespectful to their time.

**Comment by damccull** [comment]
> how has support for this mouse not been officially added yet? IT'S OVER TWO YEARS OLD

Feel free to send in a PR.

**Comment by shaqaruden** [comment]
> > how has support for this mouse not been officially added yet? IT'S OVER TWO YEARS OLD
> 
> This is an open source project run by volunteers. Having expectations with no money attached to them is quite disrespectful to contributors' time.

What I am saying is this issue has been open for 2 years and no one else in here has figured out this one Logitech G mouse and PR'd support?

**Comment by emiliopedrollo** [comment]
This is not only messing with free time of others but with their wallets as well.

It is not unheard of people bricking their devices trying to reverse engineering them. I think I could maybe crack that, but I don't want to risk brick mine. Others that could definitely do it maybe don't have this specific device and are not willing to buy such expensive mouse only to add support to it.

**Comment by damccull** [comment]
Reverse engineering is a career field. It's real hard. For digital, you have to analyze literal 1's and 0's to figure out patterns, discover what those patterns do, then see if you can replicate them. And that's all without dealing with encryption or compression.

**Comment by arnauddupuis** [comment]
> What I am saying is this issue has been open for 2 years and no one else in here has figured out this one Logitech G mouse and PR'd support?

And what have you done to make it progress in these 2 years? Have you learned any C to help the volunteers who run the project? Have you sent any PR? Have you contributed any data or tested things that others have discovered? Have you done anything except being an entitled brat? No offense, of course...

**Comment by Yeeler** [comment]
I just tried and it didn't work for me. The problems are just like @emiliopedrollo listed. 
@staticssleever668  It was too late. I did it already when I saw your comments.

**Comment by Baiyah7o** [comment]
Hi,
I can also confirm that I have `ratbagctl` and Piper to detect and/or configure my mouse, both wired and wireless. Using @drdaeman patch together with @MOVZX fork https://github.com/MOVZX/libratbag. 
```
$ ratbagctl list
screaming-chipmunk:  Logitech G502 X PLUS            
hooting-chinchilla:  Logitech USB Receiver  
```
```
$ ratbagctl "Logitech G502 X PLUS" info
screaming-chipmunk - Logitech G502 X PLUS
             Model: usb:046d:c095:0
       Device Type: Mouse
 Number of Buttons: 11
    Number of Leds: 1
Number of Profiles: 5
Profile 0:
  Name: n/a
  Report Rate: 1000Hz
  Resolutions:
    0: 100dpi
    1: 300dpi (active) (default)
    2: 600dpi
    3: 1200dpi
    4: 1600dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 4'
  Button: 4 is mapped to 'resolution-alternate'
  Button: 5 is mapped to 'button 5'
  Button: 6 is mapped to 'wheel-left'
  Button: 7 is mapped to 'wheel-right'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'resolution-up'
  Button: 10 is mapped to 'resolution-down'
  LED: 0, depth: rgb, mode: on, color: 33d17a
Profile 1:
  Name: n/a
  Report Rate: 500Hz
  Resolutions:
    0: 800dpi
    1: 1200dpi
    2: 1600dpi (active) (default)
    3: 2400dpi
    4: 3200dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 4'
  Button: 4 is mapped to 'second-mode'
  Button: 5 is mapped to 'button 5'
  Button: 6 is mapped to 'wheel-left'
  Button: 7 is mapped to 'wheel-right'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'resolution-up'
  Button: 10 is mapped to 'resolution-down'
  LED: 0, depth: rgb, mode: on, color: 33d17a
Profile 2: (disabled) (active)
Profile 3: (disabled)
Profile 4: (disabled)
```

```
$ ratbagctl "Logitech USB Receiver" info

hooting-chinchilla - Logitech USB Receiver
             Model: usb:046d:c547:0
       Device Type: Mouse
 Number of Buttons: 11
    Number of Leds: 1
Number of Profiles: 5
Profile 0:
  Name: n/a
  Report Rate: 1000Hz
  Resolutions:
    0: 100dpi
    1: 300dpi (active) (default)
    2: 600dpi
    3: 1200dpi
    4: 1600dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 4'
  Button: 4 is mapped to 'resolution-alternate'
  Button: 5 is mapped to 'button 5'
  Button: 6 is mapped to 'wheel-left'
  Button: 7 is mapped to 'wheel-right'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'resolution-up'
  Button: 10 is mapped to 'resolution-down'
  LED: 0, depth: rgb, mode: on, color: 33d17a
Profile 1: (active)
  Name: n/a
  Report Rate: 500Hz
  Resolutions:
    0: 800dpi
    1: 1200dpi
    2: 1600dpi (active) (default)
    3: 2400dpi
    4: 3200dpi
  Button: 0 is mapped to 'button 1'
  Button: 1 is mapped to 'button 2'
  Button: 2 is mapped to 'button 3'
  Button: 3 is mapped to 'button 4'
  Button: 4 is mapped to 'second-mode'
  Button: 5 is mapped to 'button 5'
  Button: 6 is mapped to 'wheel-left'
  Button: 7 is mapped to 'wheel-right'
  Button: 8 is mapped to 'profile-cycle-up'
  Button: 9 is mapped to 'resolution-up'
  Button: 10 is mapped to 'resolution-down'
  LED: 0, depth: rgb, mode: on, color: 33d17a
Profile 2: (disabled)
Profile 3: (disabled)
Profile 4: (disabled)
```
As stated before by @emiliopedrollo settings are read inaccurate but changes including LED commands can be applied to the mouse by `ratbagctl`  or Piper.

**Comment by LarveyOfficial** [comment]
Was able to confirm that the patch provided by @drdaeman worked perfectly with my G502 X Plus, using the powerplay wireless mat. I was able to remap all the buttons. Piper also kind of worked, with the exception of button 8, 9, and 10 which I could just apply via ratbagctl.

RGB did not work, stated it was OFF, but I just applied that via Ghub on windows and saved it to memory.

**Comment by UNIKO585** [comment]
Any updates on this ....work around, waiting on patch , update?
Thanks

**Comment by nuke-haus** [comment]
As of july 9, 2026, i am still getting issues with my G502 X Plus not being detected.

I added a new device entry as others have done, and i get the same verbose errors noted in https://github.com/libratbag/libratbag/issues/1344#issuecomment-1254004585

Is this planned to be fixed any time soon?
